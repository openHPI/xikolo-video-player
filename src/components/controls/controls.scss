@use "../../utils/variables.scss";

.controls {
  background-color: var(--vp-gutter-color);
  display: flex;
  flex: 1 0 auto;
  flex-flow: row wrap;

  &--fullscreen-mode {
    @media only screen and (min-width: 992px) {
      .controls__subtitle-row {
        font-size: 2.5em;
      }
      .controls__toolbar {
        .controls__playback-rate {
          &__menu {
            // Needed for matching the bottom position of the settings menu.
            margin-bottom: -3px;
            bottom: 60px;
          }

          &__button {
            font-size: 1em;
          }
        }

        .controls__time {
          font-size: 1.2em;
        }

        svg,
        .svg {
          height: 2em;
          width: 2em;
        }
      }
    }
  }

  @media only screen and (min-width: 376px) {
    .controls__toolbar {
      button {
        padding: 0.6em 0.7em;
      }
    }
  }

  @media only screen and (min-width: 992px) {
    .controls__toolbar .controls__slider--volume {
      display: none;
    }
  }

  @media only screen and (max-width: 580px) {
    .controls__subtitle {
      bottom: 40px;

      &-row {
        font-size: 1.3em;
        line-height: 1em;
        padding: 5px;
      }
    }
    .controls__toolbar {
      .controls__slider--volume {
        display: none;
      }

      .controls__time {
        padding: inherit;
      }
    }
  }

  // Reduces the font size for the current time display, so it does not wrap into two lines
  @media only screen and (max-width: 590px) {
    .controls__toolbar {
      .controls__time {
        font-size: 0.75em;
      }
    }
  }
}

.controls__playback-rate {
  position: relative;

  &__menu {
    position: absolute;
    z-index: variables.$z-index-l;
    // Needed for matching the bottom position of the settings menu.
    margin-bottom: -1px;
    bottom: 50px;
    right: 50%;
    transform: translateX(50%);
    background: var(--vp-menu-color);
    font-size: 0.9em;
    line-height: 0.9em;
    padding: 0.2em;
    visibility: hidden;
    opacity: 0;
    transition:
      visibility 0.3s linear,
      opacity 0.3s linear;

    &--open {
      visibility: visible;
      opacity: 1;
    }
  }

  &__button {
    font-size: 0.9em;
    line-height: 0.9em;
    width: 40px;
    box-sizing: content-box;
    text-align: left;
    position: relative;
    text-indent: 1em;

    &:hover {
      background: rgba(255, 255, 255, 0.2);
    }

    &--current {
      font-weight: 600;
      &:before {
        content: "\2022";
        color: var(--vp-main-color);
        position: absolute;
        left: -0.1em;
      }
    }
  }
}

.controls__subtitle {
  position: absolute;
  left: 0px;
  right: 0px;
  z-index: variables.$z-index-l;
  bottom: 60px;
  align-self: center;
  text-align: center;
  max-width: 90%;
  margin: auto;

  &-row {
    background: rgba(50, 50, 50, 0.7);
    padding: 10px;
    margin: 0.2em;
    border-radius: 10px;
    display: inline-block;
    white-space: pre-wrap;
    font-size: 1.5em;
    line-height: 1.3em;
  }
}

.controls__slider {
  flex: 1 1 100%;
  appearance: none;
  overflow: hidden;
  height: var(--vp-control-slider-height);
  background: transparent;
  color: var(--vp-slider-color);
  margin: 0;
  border-top: 1px solid black;
  border-top-color: var(--vp-bg-color);
  border-bottom: 1px solid black;
  border-bottom-color: var(--vp-bg-color);
  cursor: pointer;

  &::-webkit-slider-thumb {
    appearance: none;
    box-shadow: -100vw 0 0 100vw currentColor;
    // Safari need a fixed width to display the slider thumb
    width: 1px;
    height: var(--vp-control-slider-height);
  }

  &::-moz-range-thumb {
    appearance: none;
    box-shadow: -100vw 0 0 100vw currentColor;
    width: 0;
    border: 0;
  }

  &--volume {
    border: 1px solid black;

    &:focus {
      border: 1px solid var(--vp-main-color);
    }
  }
}

.controls__toolbar {
  margin: 1px;
  color: inherit;

  display: flex;
  flex: 1 1 100%;
  flex-flow: row nowrap;
  align-items: center;

  .controls-volume {
    display: flex;
    flex-flow: row nowrap;
    align-items: center;

    &:hover {
      @media only screen and (min-width: 376px) {
        .controls__slider--volume {
          display: block;
        }
      }
    }
  }

  .controls__time {
    flex-grow: 1;
    font-size: 0.9em;
    padding: 0.6em 0.7em;
  }

  .controls__settings-icon {
    display: block;
    overflow: hidden;
    transition-duration: 0.4s;
    transition-property: transform;

    &--open,
    &:hover {
      transform: rotate(90deg);
    }
  }

  .controls__button-icon {
    display: block;
    overflow: hidden;
    transition: box-shadow 0.3s ease-in-out;

    &--active,
    &:hover {
      box-shadow: 0px 2px 1px -1px rgba(231, 166, 26, 0.72);
    }
  }

  button {
    border: 0;
    padding: 0.6em 0.5em;
    margin: 0;
    background: transparent;
    color: inherit;
    cursor: pointer;

    &:hover {
      & .controls__settings-icon {
        transform: rotate(90deg);
      }
      & .controls__subtitle-icon {
        box-shadow: 0px 2px 1px -1px rgba(231, 166, 26, 0.72);
      }
    }
  }

  svg,
  .svg {
    display: block;
    height: 1.5em;
    width: 1.5em;
    fill: currentColor;
    color: inherit;
  }
}
