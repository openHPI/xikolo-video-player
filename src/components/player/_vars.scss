:host {
  --vp-slider-color: var(--vps-slider-color, orange);
  --vp-gutter-color: var(--vps-gutter-color, #353535);
  --vp-main-color: var(--vps-main-color, white);
  --vp-bg-color: var(--vps-bg-color, black);
  --vp-bg-color-rgb: var(--vps-bg-color-rgb, 0, 0, 0);
  --vp-menu-color-rgb: var(--vps-menu-color-rgb, 50, 50, 50);
  --vp-control-slider-height: var(--vps-control-slider-height, 0.5em);
}
