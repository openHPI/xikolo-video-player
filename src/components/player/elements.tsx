import { FunctionalComponent, h } from '@stencil/core';
import { loading as loadingIcon } from '../../utils/icon';
import { Status } from '../../utils/status';

interface LoadingProps {
  status: Status;
}

export const Loading: FunctionalComponent<LoadingProps> = ({
  status: { loading, fullscreen },
}) => {
  return (
    loading && (
      <div class={fullscreen ? 'loading loading--fullscreen-mode' : 'loading'}>
        <span class="loading__spinner" innerHTML={loadingIcon} />
      </div>
    )
  );
};
