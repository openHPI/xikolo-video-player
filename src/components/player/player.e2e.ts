import { newE2EPage } from '@stencil/core/testing';

describe('xm-player', () => {
  it('renders', async () => {
    const page = await newE2EPage();

    await page.setContent('<xm-player></xm-player>');
    const element = await page.find('xm-player');
    expect(element).toHaveClass('hydrated');
  });
});
