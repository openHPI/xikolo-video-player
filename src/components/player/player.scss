@import "../../utils/variables.scss";

:host {
  display: block;
  color: white;
  color: var(--vp-main-color);
  background: rgba(0, 0, 0, 0.75);
  background: rgba(var(--vp-bg-color-rgb), 0.75);
  border: 1px solid black;
  border-color: var(--vp-bg-color);

  font-family:
    // Basic web for OS and iOS (ensures monospaced digits)
    "Helvetica Neue",
    // Safari for OS X and iOS (San Francisco)
    -apple-system,
    // Chrome < 56 for OS X (San Francisco)
    BlinkMacSystemFont,
    // Windows
    "Segoe UI",
    // Android
    "Roboto",
    // Linux
    "Ubuntu",
    "Cantarell",
    // Basic web fallback
    Helvetica,
    Arial, sans-serif;
}

.player {
  display: flex;
  flex: 1 0 auto;
  flex-flow: row nowrap;
  flex-direction: column;
  position: relative;

  width: 100%;
  height: 100%;

  xm-screen {
    flex-grow: 1;

    ::slotted(*) {
      height: 100%;
    }
  }
}

@import "./vars";

.loading {
  animation: 0s linear 750ms forwards delayedShow;
  visibility: hidden;

  position: absolute;
  inset: 0;
  margin-bottom: 46px;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  background-color: rgba(0, 0, 0, 0.5);

  z-index: $z-index-m;

  &--fullscreen-mode {
    margin-bottom: 55px;
  }

  &__spinner {
    fill: currentColor;
    position: relative;
    width: 70px;
    height: 70px;
    animation: spinner 1.3s linear infinite;
  }

  @keyframes spinner {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

  @keyframes delayedShow {
    to {
      visibility: visible;
    }
  }
}
