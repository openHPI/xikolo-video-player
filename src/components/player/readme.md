# video-player



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description | Type     | Default                |
| -------- | --------- | ----------- | -------- | ---------------------- |
| `volume` | `volume`  |             | `number` | `defaultStatus.volume` |


## Methods

### `mute() => Promise<void>`



#### Returns

Type: `Promise<void>`



### `pause() => Promise<void>`



#### Returns

Type: `Promise<void>`



### `play() => Promise<void>`



#### Returns

Type: `Promise<void>`



### `seek(seconds: number) => Promise<void>`



#### Returns

Type: `Promise<void>`



### `unmute() => Promise<void>`



#### Returns

Type: `Promise<void>`




## Dependencies

### Depends on

- [xm-screen](../screen)
- [xm-controls](../controls)

### Graph
```mermaid
graph TD;
  xm-player --> xm-screen
  xm-player --> xm-controls
  xm-controls --> xm-settings-menu
  style xm-player fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
