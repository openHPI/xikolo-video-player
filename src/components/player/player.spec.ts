import { newSpecPage } from '@stencil/core/testing';
import { Player } from './player';
import { Controls } from '../controls/controls';
import { defaultStatus, Mode } from '../../utils/status';
import { WebVTT } from '../../utils/webVTT';

describe('xm-player default functionality', () => {
  let player: Player;

  beforeEach(async () => {
    player = new Player();
    // Mock primary video slot, so initialization does not crash
    // @ts-ignore
    player.primary = player.el.appendChild(document.createElement('div'));
  });

  /*
   * Tests for public API
   */
  it('play will set the mode to playing', async () => {
    expect(player.getStatus().mode).toEqual(Mode.PAUSED);
    await player.play();
    expect(player.getStatus().mode).toEqual(Mode.PLAYING);
  });

  it('pause will set the mode to pause ', async () => {
    expect(player.getStatus().mode).toEqual(Mode.PAUSED);
    await player.play();
    await player.pause();
    expect(player.getStatus().mode).toEqual(Mode.PAUSED);
  });

  it('seeking on a paused player will not trigger playing mode', async () => {
    expect(player.getStatus().mode).toEqual(Mode.PAUSED);
    await player.seek(10);
    expect(player.getStatus().mode).toEqual(Mode.PAUSED);
  });

  it('seeking will trigger playing, if the player is in playing mode', async () => {
    await player.play();
    await player.seek(10);
    expect(player.getStatus().mode).toEqual(Mode.PLAYING);
  });
});

describe('xm-player with props', () => {
  let page, controls;

  it('should build', () => {
    // Vimeo API is not working here, so we can not create/test the video component.
    expect(new Player()).toBeTruthy();
    expect(new Controls()).toBeTruthy();
  });

  beforeEach(async () => {
    page = await newSpecPage({
      html: `
      <xm-player lang="en" volume="0.3" playbackrate="2">
        <xm-video slot="primary" src="340196868"></xm-video>
        <xm-video slot="secondary" src="340196788"></xm-video>
      </xm-player>
    `,
      components: [Player, Controls],
    });
    // We need to render all child components manually for access to their shadow doms!
    controls = page.root.shadowRoot.querySelector('xm-controls');
    const controlsInstance = new Controls();
    controlsInstance.status = defaultStatus;
    controls.innerHTML = controlsInstance.render();
    await page.waitForChanges();
    expect(controls.shadowRoot).toBeTruthy();
  });

  it('should render', async () => {
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root.shadowRoot.querySelector('.player')).toBeTruthy();
    expect(page.root.shadowRoot.querySelector('xm-controls')).toBeTruthy();
    expect(page.root.getAttribute('lang')).toBe('en');
    expect(page.root.getAttribute('volume')).toBe('0.3');
    expect(page.root.getAttribute('playbackrate')).toBe('2');
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('should render with ui language change', async () => {
    page.rootInstance.lang = 'de';
    await page.waitForChanges();
    expect(page.rootInstance.status.language).toBe('de');
    expect(page.root).toMatchSnapshot();
  });

  it('should render with playback rate change', async () => {
    page.rootInstance.playbackrate = 1.75;
    await page.waitForChanges();
    expect(page.rootInstance.status.settings.playbackRate).toBe(1.75);
    expect(page.root).toMatchSnapshot();
  });

  it('should render with volume change', async () => {
    expect(controls.shadowRoot.querySelector('.controls-volume')).toBeTruthy();
    let button = controls.shadowRoot.querySelector('.controls__mute');
    expect(button).not.toBeNull();
    page.rootInstance.volume = 0;
    await page.waitForChanges();
    expect(page.rootInstance.status.volume).toBe(0);
    expect(page.rootInstance.status.muted).toBe(true);
    expect(button).toHaveClass('controls__unmute');
  });
});

describe('xm-player with subtitle props', () => {
  let page, controls, controlsInstance;

  beforeEach(async () => {
    page = await newSpecPage({
      html: `
      <xm-player lang="en" showsubtitle="true">
        <xm-video slot="primary" src="340196868"></xm-video>
        <xm-video slot="secondary" src="340196788"></xm-video>
        <xm-text-track language="de" src="../../static/de.vtt" label="Deutsch" default></xm-text-track>
        <xm-text-track language="en" src="../../static/en.vtt" label="English"></xm-text-track>
      </xm-player>
    `,
      components: [Player, Controls],
    });
    // We need to render all child components manually for acces to their shadow doms!
    controls = page.root.shadowRoot.querySelector('xm-controls');
    controlsInstance = new Controls();
    controlsInstance.status = page.rootInstance.status;
    controlsInstance.textTracks = page.rootInstance.textTracks;
    controls.innerHTML = controlsInstance.render();
    await page.waitForChanges();
    expect(controls.shadowRoot).toBeTruthy();
  });

  it('should render with default text track and an active subtitle button', async () => {
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root.shadowRoot.querySelector('.player')).toBeTruthy();
    expect(page.root.shadowRoot.querySelector('xm-controls')).toBeTruthy();
    expect(page.root.getAttribute('showsubtitle')).toBe('true');
    expect(
      controls.shadowRoot.querySelector('.controls__button')
    ).not.toBeTruthy();
    // Mock a loaded textTrack file
    const vtt: WebVTT = {
      meta: {
        language: 'de',
        label: 'Deutsch',
      },
      cues: [
        {
          end: 1,
          identifier: '1',
          start: 1,
          styles: '',
          text: 'test',
        },
      ],
      strict: true,
      valid: true,
      index: 0,
    };
    // Trigger the texttrack:loaded event
    let ttlEvent = new CustomEvent('texttrack:loaded', {
      detail: {
        webVTT: vtt,
        total: 1,
        isDefault: true,
      },
    });
    page.root.dispatchEvent(ttlEvent);
    await page.waitForChanges();
    expect(page.rootInstance.status.subtitle.language).toBe('de');
    controlsInstance.status = page.rootInstance.status;
    controlsInstance.textTracks = page.rootInstance.textTracks;
    controls.innerHTML = controlsInstance.render();
    await page.waitForChanges();
    expect(controls.shadowRoot).toBeTruthy();
    // Finally test the rendered output
    expect(controls.shadowRoot.querySelector('.controls__button')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });
});
