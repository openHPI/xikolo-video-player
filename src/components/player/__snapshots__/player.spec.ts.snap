// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xm-player with props should render 1`] = `
<xm-player lang="en" volume="0.3">
  <mock:shadow-root>
    <div class="player">
      <xm-screen>
        <slot name="primary" slot="primary"></slot>
        <slot name="secondary" slot="secondary"></slot>
      </xm-screen>
      <xm-controls>
        <mock:shadow-root>
          <div class="controls">
            <xm-settings-menu></xm-settings-menu>
            <input autocomplete="off" class="controls__slider" max="0" min="0" part="slider" step="any" type="range" value="0">
            <div class="controls__toolbar">
              <button title="Play">
                <span class="svg">
                  <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                    <path d="M24.267 14.867l-4.867-2.867c-1.067-0.6-2.867-1.667-3.933-2.333l-5.067-2.867c-1.067-0.667-2.133-0.133-2.133 1.133v16.067c0 1.267 1.067 1.8 2.133 1.133l4.933-2.867c1.067-0.6 2.933-1.667 4-2.333l4.867-2.867c1.133-0.533 1.133-1.6 0.067-2.2z"></path>
                  </svg>
                </span>
              </button>
              <div class="controls-volume">
                <button class="controls__unmute" title="Unmute">
                  <span class="svg">
                    <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                      <path d="M22.133 21.8l-2.533-2.533 3.267-3.267-3.267-3.267 2.533-2.533 3.267 3.267 3.267-3.267 2.533 2.533-3.333 3.267 3.333 3.267-2.533 2.533-3.267-3.267zM16.733 4.133c-0.533 0-1.067 0.267-1.533 0.667l-5.2 4.533v13.4l5.2 4.533c0.267 0.267 0.6 0.467 0.933 0.533 0.267 0.067 0.533 0.067 0.733 0 0.267-0.067 0.467-0.133 0.733-0.267 0.467-0.267 0.8-0.733 0.8-1.333v-20.533c0.067-0.867-0.8-1.467-1.667-1.533v0zM8.333 9.333v13.333h-3.8c-0.667 0-1.2-0.2-1.6-0.6s-0.533-1-0.533-1.6v-9c0-0.667 0.2-1.2 0.667-1.6s1.067-0.533 1.733-0.533h3.533z"></path>
                    </svg>
                  </span>
                </button>
                <input class="controls__slider controls__slider--volume" max="1" min="0" step="0.1" type="range" value="0">
              </div>
              <span class="controls__time" title="Time left of 00:00s">
                00:00 / 00:00
              </span>
              <div class="controls__playback-rate">
                <div class="controls__playback-rate__menu">
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button controls__playback-rate__button--current"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                </div>
                <button title="Playback rate">
                  <span class="controls__shortcut-icon svg">
                    <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                      <path d="M6.857 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM10.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM17.929 21.161l1.804-6.821q0.107-0.464-0.134-0.866t-0.688-0.527-0.857 0.116-0.536 0.705l-1.804 6.821q-1.071 0.089-1.911 0.777t-1.125 1.759q-0.357 1.375 0.357 2.607t2.089 1.589 2.607-0.357 1.589-2.089q0.286-1.071-0.107-2.089t-1.286-1.625zM29.714 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM18.286 9.143q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM26.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM32 20.571q0 4.661-2.518 8.625-0.339 0.518-0.964 0.518h-25.036q-0.625 0-0.964-0.518-2.518-3.946-2.518-8.625 0-3.25 1.268-6.214t3.411-5.107 5.107-3.411 6.214-1.268 6.214 1.268 5.107 3.411 3.411 5.107 1.268 6.214z"></path>
                    </svg>
                  </span>
                </button>
              </div>
              <button title="Settings">
                <span class="controls__settings-icon svg">
                  <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                    <path d="M18.286 16q0-1.893-1.339-3.232t-3.232-1.339-3.232 1.339-1.339 3.232 1.339 3.232 3.232 1.339 3.232-1.339 1.339-3.232zM27.429 14.054v3.964q0 0.214-0.143 0.411t-0.357 0.232l-3.304 0.5q-0.339 0.964-0.696 1.625 0.625 0.893 1.911 2.464 0.179 0.214 0.179 0.446t-0.161 0.411q-0.482 0.661-1.768 1.929t-1.679 1.268q-0.214 0-0.464-0.161l-2.464-1.929q-0.786 0.411-1.625 0.679-0.286 2.429-0.518 3.321-0.125 0.5-0.643 0.5h-3.964q-0.25 0-0.438-0.152t-0.205-0.384l-0.5-3.286q-0.875-0.286-1.607-0.661l-2.518 1.911q-0.179 0.161-0.446 0.161-0.25 0-0.446-0.196-2.25-2.036-2.946-3-0.125-0.179-0.125-0.411 0-0.214 0.143-0.411 0.268-0.375 0.911-1.188t0.964-1.259q-0.482-0.893-0.732-1.768l-3.268-0.482q-0.232-0.036-0.375-0.223t-0.143-0.42v-3.964q0-0.214 0.143-0.411t0.339-0.232l3.321-0.5q0.25-0.821 0.696-1.643-0.714-1.018-1.911-2.464-0.179-0.214-0.179-0.429 0-0.179 0.161-0.411 0.464-0.643 1.759-1.92t1.688-1.277q0.232 0 0.464 0.179l2.464 1.911q0.786-0.411 1.625-0.679 0.286-2.429 0.518-3.321 0.125-0.5 0.643-0.5h3.964q0.25 0 0.438 0.152t0.205 0.384l0.5 3.286q0.875 0.286 1.607 0.661l2.536-1.911q0.161-0.161 0.429-0.161 0.232 0 0.446 0.179 2.304 2.125 2.946 3.036 0.125 0.143 0.125 0.393 0 0.214-0.143 0.411-0.268 0.375-0.911 1.188t-0.964 1.259q0.464 0.893 0.732 1.75l3.268 0.5q0.232 0.036 0.375 0.223t0.143 0.42z"></path>
                  </svg>
                </span>
              </button>
              <button title="Enter full screen">
                <span class="svg">
                  <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                    <path d="M10.8 18.867l-5 5-3-2.933v8.267h8.267l-2.933-3 5-5zM20.933 2.8l2.933 3-5 5 2.333 2.333 5-5 3 2.933v-8.267zM10.8 13.133l2.333-2.333-5-5 2.933-3h-8.267v8.267l3-2.933zM21.2 18.867l-2.333 2.333 5 5-2.933 3h8.267v-8.267l-3 2.933z"></path>
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </mock:shadow-root>
      </xm-controls>
    </div>
  </mock:shadow-root>
  <xm-video slot="primary" src="340196868"></xm-video>
  <xm-video slot="secondary" src="340196788"></xm-video>
</xm-player>
`;

exports[`xm-player with props should render with volume 1`] = `
<xm-player lang="en" volume="0.3">
  <mock:shadow-root>
    <div class="player">
      <xm-screen>
        <slot name="primary" slot="primary"></slot>
        <slot name="secondary" slot="secondary"></slot>
      </xm-screen>
      <xm-controls>
        <mock:shadow-root>
          <div class="controls">
            <xm-settings-menu></xm-settings-menu>
            <input autocomplete="off" class="controls__slider" max="0" min="0" part="slider" step="any" type="range" value="0">
            <div class="controls__toolbar">
              <button title="Play">
                <span class="svg">
                  <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                    <path d="M24.267 14.867l-4.867-2.867c-1.067-0.6-2.867-1.667-3.933-2.333l-5.067-2.867c-1.067-0.667-2.133-0.133-2.133 1.133v16.067c0 1.267 1.067 1.8 2.133 1.133l4.933-2.867c1.067-0.6 2.933-1.667 4-2.333l4.867-2.867c1.133-0.533 1.133-1.6 0.067-2.2z"></path>
                  </svg>
                </span>
              </button>
              <div class="controls-volume">
                <button class="controls__unmute" title="Unmute">
                  <span class="svg">
                    <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                      <path d="M22.133 21.8l-2.533-2.533 3.267-3.267-3.267-3.267 2.533-2.533 3.267 3.267 3.267-3.267 2.533 2.533-3.333 3.267 3.333 3.267-2.533 2.533-3.267-3.267zM16.733 4.133c-0.533 0-1.067 0.267-1.533 0.667l-5.2 4.533v13.4l5.2 4.533c0.267 0.267 0.6 0.467 0.933 0.533 0.267 0.067 0.533 0.067 0.733 0 0.267-0.067 0.467-0.133 0.733-0.267 0.467-0.267 0.8-0.733 0.8-1.333v-20.533c0.067-0.867-0.8-1.467-1.667-1.533v0zM8.333 9.333v13.333h-3.8c-0.667 0-1.2-0.2-1.6-0.6s-0.533-1-0.533-1.6v-9c0-0.667 0.2-1.2 0.667-1.6s1.067-0.533 1.733-0.533h3.533z"></path>
                    </svg>
                  </span>
                </button>
                <input class="controls__slider controls__slider--volume" max="1" min="0" step="0.1" type="range" value="0">
              </div>
              <span class="controls__time" title="Time left of 00:00s">
                00:00 / 00:00
              </span>
              <div class="controls__playback-rate">
                <div class="controls__playback-rate__menu">
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button controls__playback-rate__button--current"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                  <button class="controls__playback-rate__button"></button>
                </div>
                <button title="Playback rate">
                  <span class="controls__shortcut-icon svg">
                    <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                      <path d="M6.857 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM10.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM17.929 21.161l1.804-6.821q0.107-0.464-0.134-0.866t-0.688-0.527-0.857 0.116-0.536 0.705l-1.804 6.821q-1.071 0.089-1.911 0.777t-1.125 1.759q-0.357 1.375 0.357 2.607t2.089 1.589 2.607-0.357 1.589-2.089q0.286-1.071-0.107-2.089t-1.286-1.625zM29.714 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM18.286 9.143q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM26.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM32 20.571q0 4.661-2.518 8.625-0.339 0.518-0.964 0.518h-25.036q-0.625 0-0.964-0.518-2.518-3.946-2.518-8.625 0-3.25 1.268-6.214t3.411-5.107 5.107-3.411 6.214-1.268 6.214 1.268 5.107 3.411 3.411 5.107 1.268 6.214z"></path>
                    </svg>
                  </span>
                </button>
              </div>
              <button title="Settings">
                <span class="controls__settings-icon svg">
                  <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                    <path d="M18.286 16q0-1.893-1.339-3.232t-3.232-1.339-3.232 1.339-1.339 3.232 1.339 3.232 3.232 1.339 3.232-1.339 1.339-3.232zM27.429 14.054v3.964q0 0.214-0.143 0.411t-0.357 0.232l-3.304 0.5q-0.339 0.964-0.696 1.625 0.625 0.893 1.911 2.464 0.179 0.214 0.179 0.446t-0.161 0.411q-0.482 0.661-1.768 1.929t-1.679 1.268q-0.214 0-0.464-0.161l-2.464-1.929q-0.786 0.411-1.625 0.679-0.286 2.429-0.518 3.321-0.125 0.5-0.643 0.5h-3.964q-0.25 0-0.438-0.152t-0.205-0.384l-0.5-3.286q-0.875-0.286-1.607-0.661l-2.518 1.911q-0.179 0.161-0.446 0.161-0.25 0-0.446-0.196-2.25-2.036-2.946-3-0.125-0.179-0.125-0.411 0-0.214 0.143-0.411 0.268-0.375 0.911-1.188t0.964-1.259q-0.482-0.893-0.732-1.768l-3.268-0.482q-0.232-0.036-0.375-0.223t-0.143-0.42v-3.964q0-0.214 0.143-0.411t0.339-0.232l3.321-0.5q0.25-0.821 0.696-1.643-0.714-1.018-1.911-2.464-0.179-0.214-0.179-0.429 0-0.179 0.161-0.411 0.464-0.643 1.759-1.92t1.688-1.277q0.232 0 0.464 0.179l2.464 1.911q0.786-0.411 1.625-0.679 0.286-2.429 0.518-3.321 0.125-0.5 0.643-0.5h3.964q0.25 0 0.438 0.152t0.205 0.384l0.5 3.286q0.875 0.286 1.607 0.661l2.536-1.911q0.161-0.161 0.429-0.161 0.232 0 0.446 0.179 2.304 2.125 2.946 3.036 0.125 0.143 0.125 0.393 0 0.214-0.143 0.411-0.268 0.375-0.911 1.188t-0.964 1.259q0.464 0.893 0.732 1.75l3.268 0.5q0.232 0.036 0.375 0.223t0.143 0.42z"></path>
                  </svg>
                </span>
              </button>
              <button title="Enter full screen">
                <span class="svg">
                  <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                    <path d="M10.8 18.867l-5 5-3-2.933v8.267h8.267l-2.933-3 5-5zM20.933 2.8l2.933 3-5 5 2.333 2.333 5-5 3 2.933v-8.267zM10.8 13.133l2.333-2.333-5-5 2.933-3h-8.267v8.267l3-2.933zM21.2 18.867l-2.333 2.333 5 5-2.933 3h8.267v-8.267l-3 2.933z"></path>
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </mock:shadow-root>
      </xm-controls>
    </div>
  </mock:shadow-root>
  <xm-video slot="primary" src="340196868"></xm-video>
  <xm-video slot="secondary" src="340196788"></xm-video>
</xm-player>
`;
